test/jdk/valhalla/valuetypes/ValueObjectMethodsTest.java

```diff
@@ -232,36 +237,52 @@ public void testToString(Object o) {
     @Test
     public void testValueRecordToString() {
         ValueRecord o = new ValueRecord(30, "thirty");
-        assertEquals(o.toString(), "ValueRecord[i=30, name=thirty]");
+        assertEquals("ValueRecord[i=30, name=thirty]", o.toString());
     }
 
-    static Stream<Arguments> hashcodeTests() {
-        Point p = new Point(0, 0);
-        Line l = new Line(0, 0, 0, 0);
-        // this is sensitive to the order of the returned fields from Class::getDeclaredFields
+    static Stream<List<Object>> hashcodeTests() {
+        Point p1 = new Point(0, 1);
+        Point p2 = new Point(0, 2);
+        Point p3 = new Point(1, 1);
+        Point p4 = new Point(2, 2);
+
+        Line l1 = new Line(0, 1, 2, 3);
+        Line l2 = new Line(9, 1, 2, 3);
+        Line l3 = new Line(0, 9, 2, 3);
+        Line l4 = new Line(0, 1, 9, 3);
+        Line l5 = new Line(0, 1, 2, 9);
+
+        Ref r1 = new Ref(p1, l1);
+        Ref r2 = new Ref(p1, l2);
+        Ref r3 = new Ref(p2, l1);
+        Ref r4 = new Ref(p2, l2);
+        Value v1 = new Value(p1, l1, r1, "s1");
+        Value v2 = new Value(p2, l1, r1, "s1");
+        Value v3 = new Value(p1, l2, r1, "s1");
+        Value v4 = new Value(p1, l1, r2, "s1");
+        Value v5 = new Value(p1, l1, r1, "s2");
+        ValueOptional vo1 = new ValueOptional(p1);
+        ValueOptional vo2 = new ValueOptional(p2);
+
+        // Each list has objects that differ from each other so the hashCodes must differ too
         return Stream.of(
-                Arguments.of(P1, hash(Point.class, 1, 2)),
-                Arguments.of(L1, hash(Line.class, new Point(1, 2), new Point(3, 4))),
-                Arguments.of(V, hash(Value.class, P1, L1, V.r, V.s)),
-                Arguments.of(new Point(0, 0), hash(Point.class, 0, 0)),
-                Arguments.of(p, hash(Point.class, 0, 0)),
-                Arguments.of(new ValueOptional(P1), hash(ValueOptional.class, P1))
+                List.of(p1, p2, p3, p4),
+                List.of(l1, l2, l3, l4, l5),
+                List.of(r1, r2, r3, r4),
+                List.of(v1, v2, v3, v4, v5),
+                List.of(vo1, vo2)
         );
     }
 
     @ParameterizedTest
     @MethodSource("hashcodeTests")
-    public void testHashCode(Object o, int hash) {
-        assertEquals(o.hashCode(), hash);
-        assertEquals(System.identityHashCode(o), hash);
-    }
+    public void testHashCode(List<Object> objects) {
+        assertTrue(objects.size() > 1, "More than one object is required: " + objects);
```

Comment:
_Nit:_ I doubt if displaying `[]` will contribute much to the diagnostics context.

```suggestion
        assertFalse(objects.isEmpty(), "More than one object is required");
```

Reply:
Every little bit helps.  
(`isEmpty` is not equivalent to `>1`)

Reply:
You're right, nevermind.